src/main/java/com/liferay/faces/util/cache/internal/ConcurrentCacheMaxCapacityLRUImpl.java

```diff
@@ -0,0 +1,218 @@
+ * recently used entry when a new value is added to the full cache. This implementation locks on writes in order to
+ * ensure that the cache cannot grow infinitely. For more details, see {@link
+ * #removeLeastRecentlyUsedCacheValuesIfNecessary(java.lang.Object)}. A simple alternative would be to use {@link
+ * Collections#synchronizedMap(java.util.Map)} on a {@link LinkedHashMap} with accessOrder set to true. However,
+ * Collections.synchronizedMap() causes the map to lock on reads which is unacceptable for a cache. For more details,
+ * see here: https://stackoverflow.com/questions/221525/how-would-you-implement-an-lru-cache-in-java
+ *
+ * @author  Kyle Stiemann
+ */
+public class ConcurrentCacheMaxCapacityLRUImpl<K, V> implements Serializable, Cache<K, V> {
+
+	// serialVersionUID
+	private static final long serialVersionUID = 6181106754606500765L;
+
+	// Private Final Data Members
+	private final ConcurrentHashMap<K, CachedValue<V>> internalCache;
+	private final Integer maxCapacity;
+
+	public ConcurrentCacheMaxCapacityLRUImpl(int initialCapacity, int maxCapacity) {
+
+		this.internalCache = new ConcurrentHashMap<K, CachedValue<V>>(initialCapacity);
+		this.maxCapacity = maxCapacity;
+	}
+
+	@Override
+	public boolean containsKey(K key) {
+		return internalCache.containsKey(key);
+	}
+
+	@Override
+	public V get(K key) {
+
+		CachedValue<V> cachedValue = internalCache.get(key);
+
+		if (cachedValue != null) {
+			return cachedValue.getValue();
+		}
+		else {
+			return null;
+		}
+	}
+
+	@Override
+	public Set<K> keySet() {
+		return internalCache.keySet();
+	}
+
+	@Override
+	public V put(K key, V value) {
+
+		removeLeastRecentlyUsedCacheValuesIfNecessary(key);
+		internalCache.put(key, new CachedValue<V>(value));
+
+		return value;
+	}
+
+	@Override
+	public V putIfAbsent(K key, V value) {
+
+		removeLeastRecentlyUsedCacheValuesIfNecessary(key);
+
+		CachedValue<V> cachedValue = internalCache.putIfAbsent(key, new CachedValue<V>(value));
+		V retValue;
+
+		if (cachedValue != null) {
+			retValue = cachedValue.getValue();
+		}
+		else {
+			retValue = value;
+		}
+
+		return retValue;
+	}
+
+	@Override
+	public V remove(K key) {
+
+		V value = null;
+		CachedValue<V> cachedValue = internalCache.remove(key);
+
+		if (cachedValue != null) {
+			value = cachedValue.getValue();
+		}
+
+		return value;
+	}
+
+	/**
+	 * This method must be called before putting values into the map. It ensures that if the map is full and we are
+	 * trying to add a new key, the least recently used value will be removed. This method must ensure that at most one
+	 * thread will remove a value at a time. Otherwise, thread A might remove a value causing threads B, C, D, E, F,...
+	 * and Z all to see that the map is not full and add their values at the same time. This would cause the map to
+	 * expand past its set max size (potentially infinitely).
+	 */
+	private void removeLeastRecentlyUsedCacheValuesIfNecessary(K key) {
```

Comment:
@stiemannkj1 Since this method only removes a single LRU value, probably better to have the method name be singular "removeLeastRecentlyUsedCacheValueIfNecessary" as well.